README.md
Mail-Outlook
============

Perl Mail::Outlook


Changes
Revision history for Perl distribution Mail-Outlook.

0.25    2016-01-16
        Removed "Further Reading" from the POD. The links were out of date
        and I don't know of any good new ones. Also removed "Future
        Enhancements" and modified the copyright.
        
0.24    2015-11-22
        Added MANIFEST.SKIP file.
        
0.23    2014-02-08
        Fixed 39711. Can now move a message from one folder to another.
        Fixed 29599. Can now list the names of the subfolders in a folder.
        Fixed 65393. Can now send mail from more than one account.
                
0.22    2013-12-16
        Fix syntax error.

0.21    2013-12-15
        Added Alien::Microsoft::Outlook as a dependency. Installation
        should now fail due to a missing dependency if Microsoft Outlook
        is not installed. If this works, I'll tidy up the tests in the
        next release.
        Altered Makefile.PL to mark this distribution as Windows only.
        
0.20    2013-12-14
        Fixed 89168 by deleting MYMETA files.
        Changed 91podcover so that it doesn't error if the require fails.
        Added 02load.t to test that the Outlook OLE constants have been
        imported. This will cause the installation to fail if Outlook
        is missing. I think that's more sensible that skipping the tests
        and installing anyway.

0.19    2013-09-07
        Last weeks release was indexed but did not pass CPAN Testers.
        Some POD was missing, 20folders.t still did not skip correctly
        if Mail::Outlook was missing, Changes file was not formatted
        correctly.
        
0.18    2013-08-31
        0.017_004 is a developer release so nothing happened. Try again.

0.17.4  2013-08-31
        0.17.1-4 are attempts to get the distribution on to CPAN and to
        get it passed by CPAN Testers. No new functionality.
        Copyright character changed to E<copy> to all three .pm files.
        20folders.t changed to skip if Outlook not installed.
        Use version numbering as in 
        http://www.dagolden.com/index.php/369/version-numbers-should-be-boring/
        so this one is 0.017_004.
        Version changed in various json and yml files. 

0.17    2013-08-24
        - Fixed RT#73880 which was a scoping problem in 20folders.t
        - Fixed RT#35856 used suggestion from davidmcb which exception that
          it should not rely on $@ which was not being set properly.

0.16	2013-07-29
        - removal of Artistic file.
        - License updated to Artistic License v2.0.
        - remove DSLIP info, as its no longer used.
        - change file dates changed to meet W3CDTF standards.
        - added META.json.
        - test file updates.
        - reworked Makefile.PL for clarity.

0.15	2009-02-10
        - patch from Alceu Rodrigues de Freitas Junior to add save() method to
          Mail::Outlook::Message, to save a new message in the Drafts folder.

0.14	2008-09-18
        - add all_folders patch (RT#27531). Thanks to Jerome Quelin.
        - fixed 30failures.t to skip properly if Test::MockObject is not 
          installed (RT#34193).
        - Apologies to all for taking so long to get this version released :(

0.13	2007-05-09
        - Updated POD for attachments and new features.
        - Added Sent & Received methods to Outlook::Message, that correspond to
          the date/time a sent mail was sent and received.
        - Fixed a long standing bug of why readonly mail wasn't returning any
          values for the accessors.
        - Added tests for read functionality. Note that as all read tests will
          read messages in your Inbox, to protect your privacy, you must set
          the environment variable PRIVATE_TESTS for them to run.
        - Devel::Cover - 84.7%

0.12	2007-05-04
        - Forgot to put back strict after testing! Ooops :(
        - Missed Cwd from the prequisities too, although it's only required
          for testing.

0.11	2007-05-03
        - Handwritten META.yml to conform to current specs.
        - Changed all instances of LICENCE to LICENSE.
        - Added META.yml test script.
        - Added LICENSE and Artistic files.
        - POD/README updates
        - upped the versions required for POD testing
        - Added 'Win32::OLE->FreeUnusedLibraries()' to the DESTROY method. Some
          people were finding that multiple instances of Outlook were being
          left open. Suggested by Fritz Zaucker
        - Hidden Attachment fix implemented thanks to a patch by Fritz Zaucker
        - Message.pm now requires File::Basename to pass the filename of the 
          attachment.
        - Finally got the failure tests to work with scary magic!
        - More tests for attachments, displaying messages and sending messages
        - Devel::Cover - 82.6%

0.10	2006-09-29
        - Added handling for attachments (basic patch by Steven Johnson)
        - Added move message and move folder functionality
        - Added delete message and delete folder functionality
        - Added more tests.
        - Devel::Cover - 63.9%

0.09    2005-08-09
        - Some items in a mail folder may actually be a MeetingItem or  a
          AppointmentItem, which don't contain the same fields as a MailItem.
          A quick fix to return undef has been used, but this is not the
          preferred solution. (Manjunath S.)

0.08    2005-03-29
        - Added better checking for Outlook in typelib list

0.07    2005-03-27
        - Reinstated strictures
        - Added better checking for Outlook being installed
        - Added POD coverage testing
        - Added DSLIP information

0.06    2005-01-26
        - POD and comment clean up

0.05    2004-01-19T13:30:16
        - Removed all Exporter references.

0.04    2003-10-14T15:48:31
        - Rewrote main module and extended it to include
          - Mail::Outlook::Folder
          - Mail::Outlook::Message
        - Now handles reading messages from
          - Inbox
          - Outbox

Makefile.PL
#!/usr/bin/perl

use strict;
use warnings;
use 5.006;

use ExtUtils::MakeMaker;

die "No support for OS" if $^O !~ m/Win/;

WriteMakefile(
    AUTHOR              => 'Barbie <barbie@cpan.org>',
    NAME                => 'Mail::Outlook',
    VERSION_FROM        => 'lib/Mail/Outlook.pm',
    ABSTRACT_FROM       => 'lib/Mail/Outlook.pm',
    NO_META             => 1,
    PREREQ_PM           => {

        # prereqs
        'File::Basename'    => '0',
	'Win32::OLE'        => '0',
	'Win32::OLE::Const' => '0',
	'Alien::Microsoft::Outlook' => '0',

        # build/test prereqs
        'Cwd'               => '0',
        'IO::File'          => '0',
        'Test::More'        => '0.70',
        'Test::Output'      => '0'

    }
);


print <<HERE;

  Please note that Microsoft Outlook is required by this distribution.


  PLEASE NOTE:
  1) You will see several Win32::OLE errors during testing. This is expected
     as we should try and test failure as well as success :)
  2) 1 message is sent during testing. If Outlook requests your acceptance,
     you may decline, as the testing will acknowledge this.
  3) Part of the testing uses your Inbox to validate whether the distribution
     can read mail. If you are uncomfortable with this, please decline any
     Outlook security pop up and those tests will be skipped.

HERE


lib/Mail/Outlook.pm
package Mail::Outlook;

use warnings;
use strict;

use vars qw($VERSION);
$VERSION = '0.24';

#----------------------------------------------------------------------------

=head1 NAME

Mail::Outlook - mail module to interface with Microsoft (R) Outlook (R).

=head1 SYNOPSIS

  # create the object
  use Mail::Outlook;
  my $outlook = new Mail::Outlook();

  # start with a folder
  my $outlook = new Mail::Outlook('Inbox');

  # use the Win32::OLE::Const definitions
  use Mail::Outlook;
  use Win32::OLE::Const 'Microsoft Outlook';
  my $outlook = new Mail::Outlook(olInbox);

  # get/set the current folder
  my $folder = $outlook->folder();
  my $folder = $outlook->folder('Inbox');

  # get the first/last/next/previous message
  my $message = $folder->first();
     $message = $folder->next();
     $message = $folder->last();
     $message = $folder->previous();

  # read the attributes of the current message
  my $text = $message->From();
     $text = $message->To();
     $text = $message->Cc();
     $text = $message->Bcc();
     $text = $message->Subject();
     $text = $message->Body();
  my @list = $message->Attach();

  # use Outlook to display the current message
  $message->display;

  # create a message for sending
  my $message = $outlook->create();
  $message->To('you@example.com');
  $message->Cc('Them <them@example.com>');
  $message->Bcc('Us <us@example.com>; anybody@example.com');
  $message->Subject('Blah Blah Blah');
  $message->Body('Yadda Yadda Yadda');
  $message->Attach(@lots_of_files);
  $message->Attach(@more_files);    # attachments are appended
  $message->Attach($one_file);      # so multiple calls are allowed
  $message->send;

  # Or use a hash
  my %hash = (
     To      => 'you@example.com',
     Cc      => 'Them <them@example.com>',
     Bcc     => 'Us <us@example.com>, anybody@example.com',
     Subject => 'Blah Blah Blah',
     Body    => 'Yadda Yadda Yadda',
  );

  my $message = $outlook->create(%hash);
  $message->display(%hash);
  $message->send(%hash);

=head1 DESCRIPTION

This module was written to overcome the problem of sending mail messages,
where Microsoft (R) Outlook (R) is the only mail application available.
However, since it's inception the module has expanded to handle a range of
Outlook mail functionality.

Note that when sending messages, the module uses the named owner of the
Outbox MAPI Folder in order to access the correct objects. Thus the From
field of a new message is predetermined, and therefore a read only property.

If using the 'Win32::OLE::Const' constants, only the following are supported:

  olFolderInbox
  olFolderOutbox
  olFolderSentMail
  olFolderDrafts
  olFolderDeletedItems

=head1 ABSTRACT

A mail module to interface with mail message accessible via
Microsoft (R) Outlook (R).

=cut

#----------------------------------------------------------------------------

#############################################################################
#Library Modules                                                            #
#############################################################################

use lib qw(./lib);

use Win32::OLE;
use Win32::OLE::Const 'Microsoft Outlook';

use Mail::Outlook::Folder;
use Mail::Outlook::Message;

#----------------------------------------------------------------------------

#############################################################################
#Interface Functions                                                        #
#############################################################################

=head1 METHODS

=head2 new()

Create a new Outlook mail object. Returns the object on success or undef on
failure. To see the last error use 'Win32::OLE->LastError();'.

=cut

sub new {
    my ($self, $foldername) = @_;

    #open the Outlook program and get a hook into it
    my $outlook;
    eval {
        $outlook = Win32::OLE->GetActiveObject('Outlook.Application')
    };
    if ($@ || !defined($outlook)) {
        $outlook = Win32::OLE->new('Outlook.Application', sub {$_[0]->Quit;})
            or return undef;
    }
    my $namespace = $outlook->GetNameSpace("MAPI")  or return undef;

    # create an attributes hash
    my $atts = {
        'outlook'   => $outlook,
        'namespace' => $namespace,
        'objfolder' => undef,
    };

    # create the object
    bless $atts, $self;

    # create a folder if required
    $atts->{objfolder} = $atts->folder($foldername) if($foldername);

    return $atts;
}

=head2 DESTROY

Free unused Win32::OLE libraries

=cut

sub DESTROY {
    Win32::OLE->FreeUnusedLibraries();
}

=head2 all_folders()

Get the list of all open mail folders in Outlook

=cut

sub all_folders {
    my ($self) = @_;
    my $folders = $self->{namespace}->Folders;
    my @folders;
    push @folders, $folders->Item($_)->Name
        for 1 .. $folders->Count;
    return @folders;
}

=head2 folder()

Gets or sets the current folder object.

=cut

sub folder {
    my ($self,$foldername) = @_;
    return $self->{objfolder}   unless($foldername);
    $self->{objfolder} = Mail::Outlook::Folder->new($self,$foldername);
}

=head2 all_accounts

Return a list of the available accounts. Each element is a hash ref with two
keys: address and account. "address" is the SMPT address, "account" is the Account
object.

=cut

sub all_accounts {
  my $self = shift;
  my $session = $self->{namespace}->Session || return;
  my $accounts = $session->Accounts || return;
  my @smtp_addresses = ();
  for ( my $i=1;$i<=$accounts->{Count};$i++ ) {
    my $item = $accounts->Item($i);
    push( @smtp_addresses, { address => $item->smtpAddress, account => $item } );  
  }
  return @smtp_addresses;  
}

=head2 create(%hash)

Creates a new message. Option hash table can be used. Returns the new message
object or undef on failure.

=cut

sub create {
    my ($self,%hash) = @_;

    my $msg = Mail::Outlook::Message->new($self->{outlook}) or return undef;
    $msg->create(%hash);
    return $msg;
}

1;

__END__

#----------------------------------------------------------------------------

=head1 FURTHER READING

If you intend to supply a patch for a bug or new feature, please visit the
following URL (and associated pages) to ensure you are using the correct
objects and methods.

http://msdn.microsoft.com/library/default.asp?url=/library/en-us/off2000/html/olobjApplication.asp

This article contains some interesting background into creating mail
messages via Outlook, although it is VB-centric.

http://www.exchangeadmin.com/Articles/Index.cfm?ArticleID=4657

=head1 FUTURE ENHANCEMENTS

A couple of items that I'd like to get working.

* X-Header support
* Send without the popups (Outlook Redemption looks possible)

=head1 NOTES

This module is intended to be used on Win32 platforms only, with Microsoft (R)
Outlook (R) installed.

  Microsoft and Outlook are registered trademarks and the copyright 1995-2003
  of Microsoft Corporation.

=head1 SEE ALSO

  Win32::OLE
  Win32::OLE::Const

=head1 BUGS, PATCHES & FIXES

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties that are not explained within the POD
documentation, please submit a bug to the RT system (see link below). However,
it would help greatly if you are able to pinpoint problems or even supply a
patch.

Fixes are dependent upon their severity and my availability. Should a fix not
be forthcoming, please feel free to (politely) remind me by sending an email
to barbie@cpan.org .

RT: L<http://rt.cpan.org/Public/Dist/Display.html?Name=Mail-Outlook>

=head1 AUTHOR

  Barbie, <barbie@cpan.org>
  for Miss Barbell Productions, <http://www.missbarbell.co.uk>

=head1 COPYRIGHT AND LICENSE

  Copyright E<copy> 2003-2013 Barbie for Miss Barbell Productions.

  This distribution is free software; you can redistribute it and/or
  modify it under the Artistic License v2.

=cut


lib/Mail/Outlook/Folder.pm
package Mail::Outlook::Folder;

use warnings;
use strict;

use vars qw($VERSION);
$VERSION = '0.24';

#----------------------------------------------------------------------------

=head1 NAME

Mail::Outlook::Folder - extension to handle Microsoft (R) Outlook (R) mail folders.

=head1 SYNOPSIS

See Mail::Outlook, as this is not meant to be used as a standalone module.

=head1 DESCRIPTION

Handles the Folder interaction with the Outlook API.

=cut

#----------------------------------------------------------------------------

#############################################################################
#Library Modules                                                            #
#############################################################################

use Win32::OLE;
use Win32::OLE::Const 'Microsoft Outlook';

use Mail::Outlook::Message;

#############################################################################
#Variables
#############################################################################

my %foldernames = (
    'Inbox'         => olFolderInbox,
    'Outbox'        => olFolderOutbox,
    'Sent Items'    => olFolderSentMail,
    'Drafts'        => olFolderDrafts,
    'Deleted Items' => olFolderDeletedItems,
);

#----------------------------------------------------------------------------

#############################################################################
#Interface Functions                                                        #
#############################################################################

=head1 METHODS

=over 4

=item new()

Create a new Outlook mail object. Returns the object on success or undef on
failure. To see the last error use 'Win32::OLE->LastError();'.

=cut

sub new {
    my ($self, $outlook, $foldername) = @_;
    my ($mailbox,$folder,$path);

    # split mailbox and path
    ($foldername,$path) = ($foldername =~ m!(.*?)/(.*)!)
        if ($foldername =~ m!/!);

    # mailbox name
    if($foldernames{$foldername}) {
        eval { $mailbox = $outlook->{namespace}->GetDefaultFolder($foldernames{$foldername}) };
        return undef    if($@);

    # mailbox constant only
    } elsif($foldername =~ /^\d+$/) {
        eval { $mailbox = $outlook->{namespace}->GetDefaultFolder($foldername) };
        return undef    if($@);

    # well if you don't know, neither do i!!!
    } else {
        return undef;
    }

    if($path) {
        # This is a bit of a hack to stop the OLE complaining when the path
        # doesn't exist in the folder tree
        my $hash;
        eval { $hash = $mailbox->Folders(); };
        my %keys = map {$_ => 1} keys %$hash;
        return undef    if($@);

        $folder = $self->_folders($mailbox, $path) || return undef; 

    }

    # create an attributes hash
    my $atts = {
        'outlook'       => $outlook,
        'foldername'    => $foldername,
        'objfolder'     => $folder || $mailbox || undef,
        'items'         => undef,
    };

    # prime the mail items collection
    $atts->{items} = $atts->{objfolder}->Items()    or return undef;

    # create the object
    bless $atts, $self;
    return $atts;
}

# Split the path (eg A/B/C) into segments call call the Folder method for each. 
# Unix style separators (/).
sub _folders {
  my ($self,$mailbox, $path)=@_;
  my @segs = split('/', $path );
  my $f;
  foreach my $s ( @segs ) {
    eval { $f = $f ? $f->Folders($s) : $mailbox->Folders($s);1;}||return undef;
    return undef if($@); 
    return undef unless( defined($f) );    
  }
  return $f;
}

=item DESTROY

Empty sub

=cut

sub DESTROY {}

=item first()

Gets the first message object in the current folder. Returns undef if no messages.

=cut

sub first {
    my $self = shift;
    my $message = $self->{items}->GetFirst();
    return  unless($message);
    Mail::Outlook::Message->new($self->{outlook},$message);
}

=item last()

Gets the last message object in the current folder. Returns undef if no messages.

=cut

sub last {
    my $self = shift;
    my $message = $self->{items}->GetLast();
    return  unless($message);
    Mail::Outlook::Message->new($self->{outlook},$message);
}

=item next()

Gets the next message object in the current folder. Returns undef if no more
messages. Must be called after a first() or last() has been intiated.

=cut

sub next {
    my $self = shift;
    my $message = $self->{items}->GetNext();
    return  unless($message);
    Mail::Outlook::Message->new($self->{outlook},$message);
}

=item previous()

Gets the previous message object in the current folder. Returns undef if no
more messages. Must be called after a first() or last() has been intiated.

=cut

sub previous {
    my $self = shift;
    my $message = $self->{items}->GetPrevious();
    return  unless($message);
    Mail::Outlook::Message->new($self->{outlook},$message);
}

=item move($message)

Move a message into this folder.

=cut

sub move {
    my ($self,$message) = @_;

    $message->{message}->Move($self->{objfolder});
    return 1;
}

=item move_folder($folder)

Move a folder into this folder.

=cut

sub move_folder {
    my ($self,$folder) = @_;

    $folder->{objfolder}->Move($self->{objfolder});
}

=item delete_folder()

Remove this folder.

=cut

sub delete_folder {
    my $self = shift;

    $self->{objfolder}->Delete();
    $self = undef;
}

=item all_folders

List the names of all folders in this folder.

=cut

sub all_folders {
  my $self = shift;
  my $folders = $self->{objfolder}->Folders || [];
  my @folder_names = ();
  my $num_folders = $folders->{Count};
  for( my $i=1; $i <= $num_folders; $i++ ) {
    push( @folder_names, $folders->Item($i)->Name );
  }
  return @folder_names;
  1;
}

=item count_items

Return the number of items in the folder.

=cut

sub count_items {
  my $self = shift;
  my $count = $self->{objfolder}->Items->Count;
}

1;

__END__

#----------------------------------------------------------------------------

=back

=head1 FURTHER READING

If you intend to supply a patch for a bug or new feature, please visit the
following URL (and associated pages) to ensure you are using the correct
objects and methods.

http://msdn.microsoft.com/library/default.asp?url=/library/en-us/off2000/html/olobjApplication.asp

This article contains some interesting background into creating mail
messages via Outlook, although it is VB-centric.

http://www.exchangeadmin.com/Articles/Index.cfm?ArticleID=4657

=head1 FUTURE ENHANCEMENTS

A couple of items that I'd like to get working.

* X-Header support
* Send without the popups (Outlook Redemption looks possible)

=head1 NOTES

This module is intended to be used on Win32 platforms only, with Microsoft (R)
Outlook (R) installed.

  Microsoft and Outlook are registered trademarks and the copyright 1995-2003
  of Microsoft Corporation.

=head1 SEE ALSO

  Win32::OLE
  Win32::OLE::Const

=head1 BUGS, PATCHES & FIXES

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties that are not explained within the POD
documentation, please submit a bug to the RT system (see link below). However,
it would help greatly if you are able to pinpoint problems or even supply a
patch.

Fixes are dependent upon their severity and my availability. Should a fix not
be forthcoming, please feel free to (politely) remind me by sending an email
to barbie@cpan.org .

RT: L<http://rt.cpan.org/Public/Dist/Display.html?Name=Mail-Outlook>

=head1 AUTHOR

  Barbie, <barbie@cpan.org>
  for Miss Barbell Productions, <http://www.missbarbell.co.uk>

=head1 COPYRIGHT AND LICENSE

  Copyright E<copy> 2003-2013 Barbie for Miss Barbell Productions.

  This distribution is free software; you can redistribute it and/or
  modify it under the Artistic License v2.

=cut
